### ISSUE TITLE
Logical order of `i` and `by`

### ISSUE THREAD

**Issue by st-pasha** [h2oai/datatable#1348]
The generic syntax of a group-by call to a Frame is:
```
DT[i, j, by()]
```
where `i` is a "row selector", `j` is a "column selector" (also known as "what to do" argument), and finally `by()` specifies how the grouping is to be done.

This generic description, however, leaves ambiguity in how the operations of grouping and selecting are ordered relative to each other. Possible interpretations are:
- The `i` row filter is applied first, and then the frame is grouped and the `j` result is computed. Thus, the whole expression conceptually is `DT[i, :][:, j, by]`.
- The frame is grouped and the `j` columns are computed, and then `i` filter is applied. Thus, the expression is conceptually equivalent to `DT[:, j, by][i, :]`.

This issue is an attempt to document my thought process about this choice, and also an attempt to solicit feedback from other developers/users.

## Performance
In the first interpretation, applying `i` to a frame is very efficient: `DT[i, :]` produces a *view* onto `DT` with `i` as the "row index". The data is not copied unnecessarily. This means that with this interpretation, the performance of `DT[i, j, by]` is exactly the same as the performance of the explicit `DT[i, :][:, j, by]`.

With the second interpretation, computing `DT[:, j, by]` first and then applying `[i, :]` is less efficient then what could potentially be achieved by doing `DT[i, j, by]`. This is because `DT[:, j, by]` has to compute the entire grouped result first, only to discard parts of it later.

This means that if the first interpretation is adopted, there is no efficient way to perform the computation if you want to treat `i` as the post-filter. At the same time, if the second interpretation is used, then the user who wants to apply `i` as a pre-filter can just write `DT[i, :][:, j, by]` and this will compute efficiently.

## Usability
The syntax should generally be intuitive. If a certain expression *looks* like it should be doing one thing but is instead doing another, then it's a bug. Because a user will not read the documentation. They'll simply presume it's a bug, and leave with the impression that `datatable` is buggy. With this understanding in mind, let's look at a few possible expressions:
```
DT[sum(f.A) > 0, sum(f.A), by(f.B)]
```
With the pre-by interpretation, the meaning of `sum`s in `i` and `j` is different: the first aggregates over the entire frame, the second within each group. With the post-by interpretation, the meaning is the same.

Another example:
```
DT[f.score == max(f.score), :, by(f.id)]
```
With post-by interpretation, this selects entries that achieved maximum `score` for every `id`. With pre-by interpretation, this selects a single globally maximum score, and then the `id` grouping is essentially useless.

When a single row number is used:
```
DT[0, :, by(f.foo)]
```
This selects the first row in each group with post-by interpretation, or the first row in `DT` with the pre-by interpretation (the `by` condition is now useless).

Similarly, to select the first and the last row within each group, with the post-by interpretation becomes as simple as:
```
DT[[0, -1], :, by(f.id)]
```

## Conclusion

Both from the performance and usability standpoint the "post-by" interpretation wins: `i` is better to treat as applied after the groupby, not before. There are many compelling examples where filtering within groups is needed, or the groups themselves must be discarded based on some criteria. 

At the same time, I could not find a single compelling reason or use-case in favor of the "pre-by" interpretation. If such use case does arise, it can always be achieved with `DT[i, :][:, j, by]` with no loss of performance or readability.

**Comment by XiaomoWu** [comment]
The R data.table adopts the **pre-by** interpretation. To solve the "post-by-in-group selection" problem, R data.table introduces a special symbol `.SD`, which means the "Subset of Dataset" for each by group. With that, the API for selecting the first row in a group would be:

```
dt[, .SD[0], by()]
```

The power of `.SD` is not limited to this special user case, it turns out to be very userful in achieving many tricky data processing requirement. Would you consider to bring a Python analogy of `.SD`?

**Comment by st-pasha** [comment]
@XiaomoWu You're right that in R data.table symbol `.SD` is indispensable. It encapsulates the "subset of data" within each group, and can be used in a variety of ways.

The simplest use case is `DT[, lapply(.SD, sum)]` to apply a certain function (`sum`)  to each column, possibly by group. In python you can do same via a list comprehension, those it's more verbose. If we want a more parsimonious syntax for this case in python, then we wouldn't introduce two separate symbols `lapply` and `SD` -- a single function would suffice. Then we could say `DT[:, allapply(sum)]` or similar.

Another use case is when `.SD` is a group of data, and we need to apply some i/j expression to that group. For this case we already have `DT[i, j, by]` syntax, since in the presence of `by` the first two selectors `i` and `j` already act within each group. In this case using `.SD` may actually be detrimental to performance, because it introduces overhead of a python call for each group, and prevents us from parallelizing the calculations.

The last use case (or maybe not last) is when you want to treat each subgroup of data as an actual `Frame`, for example saving it to disk, or passing to a regression or ML model, etc. This use-case we currently do not support, but eventually will. Note that this case is not without its own drawbacks: in order to allow the user to "do anything" with `.SD` we have to generally call a user-defined function and pass `.SD` as the parameter. But this would prevent us from parallelizing the computations (unless perhaps if we go through multi-processing).
